src/retry.ts

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { errorMessage as parseErrorMessage } from './errors';
+import { sleep } from './time';
+import { clamp } from './util';
+
+/**
+ * RetryOptions are the config options for the withRetry function.
+ */
+export interface RetryOptions {
+  /**
+   * attempts is the number of _retry_ attempts. This must be zero or a positive number.
+   */
+  attempts?: number;
+
+  /**
+   * backoff is the starting backoff time in milliseconds
+   */
+  backoff?: number;
+
+  /**
+   * backoffLimit is the maximum backoff time in milliseconds
+   */
+  backoffLimit?: number;
+}
+
+/**
+ * withRetry a simple implementation of a retry mechanism with fibonacci backoff
+ *
+ * @param fn. A function to retry on failure
+ * @param opts. The retry options
+ * @throws {Error}
+ */
+export async function withRetry<T>(fn: () => Promise<T>, opts?: RetryOptions): Promise<T> {
+  const DEFAULT_ATTEMPTS = 3;
```

Comment:
Should we move these constants outside the function body so they are only declared on load?